elasticsearch-transport/lib/elasticsearch/transport/transport/sniffer.rb

```diff
@@ -28,21 +28,23 @@ def initialize(transport)
         # @raise  [SnifferTimeoutError]
         #
         def hosts
-          Timeout::timeout(timeout, SnifferTimeoutError) do
-            nodes = transport.perform_request('GET', '_nodes/http').body
-            hosts = nodes['nodes'].map do |id,info|
-              addr_str = info["#{transport.protocol}_address"].to_s
-              matches = addr_str.match(ES1_RE_URL) || addr_str.match(ES2_RE_URL)
-              if matches
-                host = matches[1].empty? ? matches[2] : matches[1]
-                port = matches[3]
-                info.merge :host => host, :port => port, :id => id
-              end
-            end.compact
+          # We use a timeout in the client itself here because using a library like
+          # 'Timeout' that uses a separate thread would not be threadsafe.
+          # It could well be that the underlying transport has reinitialized itself in
+          # another thread while this request is still pending.
+          nodes = transport.perform_request('GET', '_nodes/http', :timeout => 1000).body
```

Comment:
Seems like Ruby's `Timeout` is indeed ["dangerous"](http://www.mikeperham.com/2015/05/08/timeout-rubys-most-dangerous-api/) and it's probably better to get rid of it. But this implementation doesn't look correctly (and the tests fail), and also uses a hardcoded value. We need to handle timeout properly here, for all adapters, Faraday (the default) and others. It might require a change in the base (abstract) class.